drivers/watchdog/bflb_wdt.c

```diff
@@ -0,0 +1,297 @@
+	WDIOF_KEEPALIVEPING |
+	WDIOF_MAGICCLOSE,
+};
+
+static const struct watchdog_ops bflb_wdt_ops = {
+	.start = bflb_wdt_start,
+	.stop = bflb_wdt_stop,
+	.ping = bflb_wdt_ping,
+	.set_timeout = bflb_wdt_set_timeout,
+	.get_timeleft = bflb_wdt_timeleft,
+};
+
+
+static int __init bflb_wdt_probe(struct platform_device *pdev)
+{
+	struct bflb_watchdog_device *bflb_wdd;
+	struct watchdog_device *wdd;
+	struct resource *res;
+	int err;
+	uint32_t reg_val;
+
+	dev_dbg(&pdev->dev, "bflb_wdt_probe started");
+
+	bflb_wdd = devm_kzalloc(&pdev->dev, sizeof(*bflb_wdd), GFP_KERNEL);
+	if (!bflb_wdd) return -ENOMEM;
+
+	res = platform_get_resource(pdev, IORESOURCE_MEM, 0);
+	bflb_wdd->regs = devm_ioremap_resource(&pdev->dev, res);
+	if (IS_ERR(bflb_wdd->regs))
+		return PTR_ERR(bflb_wdd->regs);
+
+	wdd = &bflb_wdd->wdd;
+
+	wdd->info = &bflb_wdt_info;
+	wdd->ops = &bflb_wdt_ops;
+
+	wdd->timeout = BFLB_DEFAULT_TIMEOUT;
+	wdd->max_timeout = BFLB_MAX_SECS;
+	wdd->min_timeout = 1;
+	wdd->parent = &pdev->dev;
+
+	watchdog_stop_on_reboot(wdd);
+	watchdog_stop_on_unregister(wdd);
+	watchdog_set_nowayout(wdd, WATCHDOG_NOWAYOUT);
+	watchdog_init_timeout(wdd, BFLB_DEFAULT_TIMEOUT, &pdev->dev);
+
+	// Setup registers
+
+	// Set to reboot on watchdog, disable until we start
+	bflb_unlock_watchdog(bflb_wdd);
+	reg_val = readl(bflb_wdd->regs + BFLB_REG_WMER);
+	reg_val &= ~BFLB_VAL_WE;
+	reg_val |= BFLB_VAL_WRIE;
+	writel(reg_val, bflb_wdd->regs + BFLB_REG_WMER);
+
+	// Set to 1K per second clock
+	reg_val = readl(bflb_wdd->regs + BFLB_REG_TCCR);
+	reg_val &= ~BFLB_MASK_CS_WDT;
+	reg_val |= BFLB_VAL_CS_1K;
+	writel(reg_val, bflb_wdd->regs + BFLB_REG_TCCR);
+
+	// Max out divider to 255, meaning actual
+	// ticks will happen 4 times a second, or
+	// every 250 ms.
+	reg_val = readl(bflb_wdd->regs + BFLB_REG_TCDR);
+	reg_val &= ~BFLB_MASK_TCDR;
+	reg_val |= BFLB_VAL_TCDR;
+	writel(reg_val, bflb_wdd->regs + BFLB_REG_TCDR);
+
+	// Set last valid timeout value
+	bflb_wdt_update_timeout_reg(wdd);
+
+	err = devm_watchdog_register_device(&pdev->dev, wdd);
+	if (err) return err;
+
+	platform_set_drvdata(pdev, bflb_wdd);
+
+	dev_info(&pdev->dev, "bflb_wdt_probe completed...");
+
+	return 0;
+}
+
+static int __exit bflb_wdt_remove(struct platform_device *pdev)
+{
+	dev_info(&pdev->dev, "bflb_wdt_remove removed...");
+
+	return 0;
+}
+
+static const struct of_device_id bflb_wdt_match[] = {
+	{
+		.compatible = "bflb,bflb808-wdt",
+	},
+	{},
+};
+
+MODULE_DEVICE_TABLE(of, bflb_wdt_match);
+
+static struct platform_driver bflb_wdt_driver = {
+	.probe = bflb_wdt_probe,
+	.remove = bflb_wdt_remove,
+	.driver = {
+		.name = "bflb_wdt",
+		.owner = THIS_MODULE,
+		.of_match_table = bflb_wdt_match,
+  },
+};
+
+
+static int __init bflb_wdt_init(void)
+{
+	int res;
+
+	res = platform_driver_register(&bflb_wdt_driver);
+	if(res)
+		pr_alert("bflb_wdt_init FAILED (%i)\n", res);
+	else
+		pr_alert("bflb_wdt_init: %s\n", "success");
+
+	return res;
+}
+
+arch_initcall(bflb_wdt_init);
```

Comment:
Please replace this function/initcall with `module_platform_driver`

Reply:
I've tried fixing this in [a8f44c8](https://github.com/arm000/linux-bl808/pull/8/commits/a8f44c89b43f6246175c6044d3184264cb493dd2) but then the wdt never probes.

I initially had this problem when developing the initial patch and got the fix from the `bflb_uart.c` driver file which uses this initialization method.

So I've reverted the fix with commit [bdf34f2](https://github.com/arm000/linux-bl808/pull/8/commits/bdf34f2b79c52ec2dce896220805ecf34c89ecd1)

I'll continue to investigate but if anyone else could try to build with and without commit [bdf34f2](https://github.com/arm000/linux-bl808/pull/8/commits/bdf34f2b79c52ec2dce896220805ecf34c89ecd1) to see if they can reproduce I'd appreciate it.

Reply:
[59f8395](https://github.com/arm000/linux-bl808/pull/8/commits/59f8395a3af2ddece4df78b5de73161a8352cd51) fixed the probing issue.